Jamesinglis/ag 59/setup docker

jeinglis opened the pull request:
I have the ADD command commented out to try an get the docker-compose volumes working. In it's current state it won't work. You need to uncomment ADD and then don't bother with the compose.

jeinglis committed:
ag-59 created Dockerfile and docker-compose.yml

jeinglis committed:
ag-59/added setup docker-compose.yml and Dockerfile still not buildign properly

jeinglis committed:
Merge branch 'master' into jamesinglis/ag-59/setup-docker

jeinglis committed:
added frozen option to yarn call

jeinglis committed:
the DockerFile is right and works if you ADD . but the docker-compose file doesn't work

ExtraGravity commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+RUN yarn install --frozen-lockfile
+
+RUN npm run build
```
This gets locked to the image every time you build, so you would have to rebuild the image every time you make changes. So this should be taken out, and just be a different way to run this file. 

Example
https://github.com/enochtsang/personal-website/blob/master/docker-build.sh

Here in the bash file, there are two ways to run the image, running or building, we'd want to do something similar.

ExtraGravity commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+WORKDIR /agriculture-core/
+
+#ADD . /agriculture-core/
```
The `ADD` command will also lock the changes to the image, the image will need to be reinitialized every time changes to the repo is made. So instead of building this with the ADD command, we should specify how to start the container with the `--volume` option.

jeinglis replied:
I think the issue is that I need to build the code in the container during the image build process. That's why the VOLUME command won't work. it doesn't matter if we reference it at run time because we need the code at the time we build the image so we can build the code within the container. The only way around this would be to find a way to build it when we run.

ExtraGravity reviewed (changes requested):
Yea I've had issues with docker-compose before, I thought this time would be different but nah screw it. Bash scripts for docker have always treated me well, let's do that instead of docker-compose.

jeinglis committed:
mods to Dockerfile

jeinglis committed:
removed the un-needed script files

jeinglis committed:
added start and dockerize scripts

jeinglis commented:
Dockerfile is done and two scripts have been written to kick of the init, build and run which are accessed through dockerize.sh.

beaesguerra commented on dockerize.sh:
```diff
@@ -0,0 +1,16 @@
+elif [ "$1" == "build" ]; then
+	docker run --rm -v $(pwd):/agriculture-core/ $IMAGE_NAME npm build
+elif [ "$1" == "run" ]; then
+    docker run --rm --volume "$(pwd):/agriculture-core/" -ti agriculture-core ./start.sh
```
I think maybe we don't need a start.sh script, we can just have 2 commands in this elif statement (one for yarn install and one for npm start). That way, people won't accidentally use the start script on it's own.. unless we want that?

ExtraGravity replied:
actually yea I think I agree with bea here, we'll just use two `docker run` commands.

beaesguerra commented on dockerize.sh:
```diff
@@ -0,0 +1,16 @@
+if [ "$1" == "init" ]; then
+	docker build -t $IMAGE_NAME $DIR
+elif [ "$1" == "build" ]; then
+	docker run --rm -v $(pwd):/agriculture-core/ $IMAGE_NAME npm build
```
I think this needs a command to install the packages before building

beaesguerra replied:
also this needs to be `npm run build` not npm build :)

beaesguerra commented on dockerize.sh:
```diff
@@ -0,0 +1,16 @@
+elif [ "$1" == "build" ]; then
+	docker run --rm -v $(pwd):/agriculture-core/ $IMAGE_NAME npm build
+elif [ "$1" == "run" ]; then
+    docker run --rm --volume "$(pwd):/agriculture-core/" -ti agriculture-core ./start.sh
```
I think we can use the IMAGE_NAME variable here too after the -ti flag

beaesguerra replied:
Might also be a good idea to add instructions to the README on how to use the dockerize script! (We should also probably consolidate on this since it should be the same commands for the 3 repos we are working on. I am following the way you've done it!)

beaesguerra commented on dockerize.sh:
```diff
@@ -0,0 +1,16 @@
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+IMAGE_NAME="boresha/agriculture-core"
+DOCKER_HOME="/agiculture-core/"
+ARG_NUM=2
```
do we need this?

ExtraGravity commented:
I agree with all of @beaesguerra's comments, I'll wait for them to be resolved before I review again.

Parasrah commented on package-lock.json:
```diff
@@ -0,0 +1,6374 @@
+{
+  "name": "agriculture-core",
```
# Requested Change

Since we are using yarn, this file is not necessary and shouldn't ever be generated. We could make a case to use npm for our dependencies instead of yarn, but we should only use one (or we have to run `yarn install && npm install` every time we make a pull request, effectively maintaining two lock files doing the same thing)
To clarify we should use `yarn add <package>` and `yarn install` for dependencies, and only use `npm run <script>`

jeinglis committed:
added readme

jeinglis committed:
made changes to dockerize.sh requested by reviewers: it no longer requires a second run script and uses yarn rather than npm

jeinglis committed:
deleted the start.sh which is no longer needed

jeinglis commented:
I addressed all of the above concerns. The I selected port 9090 to expose as a place holder this may not be the final port selected for the APIs to communicate over

ExtraGravity commented:
Agreed, we should probably try to pick something early that won't conflict with other services. 9090 is very common.

jeinglis committed:
removed the package-lock.json at brads request

ExtraGravity commented on dockerize.sh:
```diff
@@ -0,0 +1,17 @@
+DOCKER_HOME="/agiculture-core/"
+
+if [ "$1" == "init" ]; then
+	docker build -t $IMAGE_NAME .
```
What was wrong with `$DIR`? It let's us run this script from other directories.

jeinglis replied:
I changed it to be consistent with the dashboard dockerize.sh script. the $DIR variable was set pointing to the pwd anyway

ExtraGravity replied:
They should both use `$DIR`, the thing I have at the top is pretty robust way of getting the dir of the script.

jeinglis replied:
Okay it has been changed to $DIR

jeinglis committed:
modified dockerize.sh init command to use  rather than .

ExtraGravity commented:
@Right2Drive please QA

Parasrah commented:
Sorry didn't realize this was ready for me

Parasrah commented:
@enochtsang @jeinglis Should the instructions for docker be added to the README?

ExtraGravity commented:
They've already been added. Steps on installing docker as a whole aren't necessary in my opinion.

Parasrah commented:
Oh I just forgot to pull my bad sorry thanks @enochtsang :D

On a separate note, is the sh script being run inside the container? Or am I expected to run it on the terminal?

Parasrah reviewed (changes requested):
Just gave it a run through and received the following error on build:
```
sudo ./dockerize.sh build
yarn install v1.3.2
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
info Lockfile not saved, no dependencies.
Done in 0.07s.
npm ERR! path /agriculture-core/package.json
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open
npm ERR! enoent ENOENT: no such file or directory, open '/agriculture-core/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent
```

Looks like it can't find the `package.json`? Same thing happens for the `run` script as well

Also, we can't run the `.sh` scripts on Windows correct? Or is there something that I'm missing about docker. Couldn't we write this `.sh` script as a set of `npm` scripts instead? That way it would work independent of the platform.

ExtraGravity commented:
Well kind of the point is that you don't need npm installed to run any of this, so making it an npm script is kind of moot.

Windows has alot of different avenues to run bash scripts too, Git Bash, docker terminal, and so on. I mean James is on windows only and he wrote it, so I think it's fine.

Parasrah commented:
Alright that makes sense, any idea why it's not finding the volume?

Parasrah commented:
@enochtsang @jeinglis  Is this ready for QA again or did I just forget to move this back in the pipeline?

beaesguerra commented:
Not really your problem now @jeinglis but I noticed that in @enochtsang 's dockerize scripts, he used the run argument for running any npm script. We don't have to fix it here now, but it might be a good idea later on to do that as well so we can run tests  in docker

ExtraGravity commented:
Yea when I was advising you guys on this docker stuff, I wasn't aware of the run scripts that npm had. It's really useful and we should be using it in our dockerize scripts, otherwise we're duplicating work.

ExtraGravity committed:
ag-59 update script and remove compose file

ExtraGravity committed:
at-59 update usage line

ExtraGravity commented on dockerize.sh:
```diff
@@ -0,0 +1,17 @@
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+IMAGE_NAME="boresha/agriculture-core"
+DOCKER_HOME="/agiculture-core/"
```
nothing was building due to this typo

ExtraGravity committed:
ag-59 fix typo and tabs to spaces

ExtraGravity committed:
ag-59 fix merge conflicts

Parasrah reviewed (approved):
Works like a charm! Awesome!
